src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,1518 @@
+#define TAG_VULNERABILITY_CWE BAD_CAST "CWE"
+#define TAG_CVRF_PRODUCT_STATUSES BAD_CAST "ProductStatuses"
+#define TAG_CVRF_PRODUCT_STATUS BAD_CAST "Status"
+#define TAG_PRODUCT_ID BAD_CAST "ProductID"
+#define ATTR_STATUS_TYPE BAD_CAST "Type"
+// Remediations
+#define TAG_REMEDIATIONS BAD_CAST "Remediations"
+#define TAG_REMEDIATION BAD_CAST "Remediation"
+#define TAG_REMEDIATION_DESC BAD_CAST "Description"
+#define TAG_REMEDIATION_URL BAD_CAST "URL"
+// namespaces
+#define CVRF_NS BAD_CAST "http://www.icasi.org/CVRF/schema/cvrf/1.1"
+#define PROD_NS BAD_CAST "http://www.icasi.org/CVRF/schema/prod/1.1"
+#define VULN_NS BAD_CAST "http://www.icasi.org/CVRF/schema/vuln/1.1"
+
+/* End of XML variable definitions
+ *
+ ***************************************************************************/
+
+/****************************************************************************
+ * Parsing functions
+ *
+ */
+
+struct cvrf_index *cvrf_index_parse_xml(struct oscap_source *index_source) {
+
+	__attribute__nonnull__(index_source);
+
+	struct cvrf_index *index = cvrf_index_new();
+	cvrf_index_set_index_file(index, oscap_source_readable_origin(index_source));
+	struct oscap_stringlist *file_list = oscap_stringlist_new();
+	struct cvrf_model *model;
+
+	FILE* file = fopen(oscap_source_get_filepath(index_source), "r");
+	char line[256];
+
+	while (fgets(line, sizeof(line), file)) {
+		printf("%s", line);
+		oscap_trim(line);
+		oscap_stringlist_add_string(file_list, line);
+	}
+	fclose(file);
+
+	struct oscap_string_iterator *iterator = oscap_stringlist_get_strings(file_list);
+	while (oscap_string_iterator_has_more(iterator)) {
+		const char *filename = oscap_string_iterator_next(iterator);
+		model = cvrf_model_parse_xml(oscap_source_new_from_file(filename));
+		if (model)
+			oscap_list_add(index->models, model);
+	}
+	oscap_string_iterator_free(iterator);
+	oscap_source_free(index_source);
+
+	return index;
+}
+
+struct cvrf_model *cvrf_model_parse_xml(struct oscap_source *source) {
+
+	__attribute__nonnull__(source);
+
+	struct cvrf_model *cvrf = NULL;
+	int rc;
+
+	struct xmlTextReaderPtr *reader = oscap_source_get_xmlTextReader(source);
+	if (!reader) {
+		oscap_source_free(source);
+		return NULL;
+	}
+
+	rc = xmlTextReaderNextNode(reader);
+	if (rc == -1) {
+		xmlFreeTextReader(reader);
+		oscap_source_free(source);
+		return NULL;
+	}
+
+	cvrf = cvrf_model_parse(reader);
+
+	xmlFreeTextReader(reader);
+	oscap_source_free(source);
+
+	return cvrf;
+}
+
+struct cvrf_model *cvrf_model_parse(xmlTextReaderPtr reader) {
+	__attribute__nonnull__(reader);
+
+	struct cvrf_model *ret = NULL;
+	struct cvrf_doc_tracking *tracking = NULL;
+	struct cvrf_doc_publisher *publisher = NULL;
+	struct cvrf_vulnerability *vulnerability = NULL;
+	struct cvrf_product_tree *tree = NULL;
+	struct cvrf_branch *branch = NULL;
+
+	if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_CVRF_DOC) &&
+	    xmlTextReaderNodeType(reader) == XML_READER_TYPE_ELEMENT) {
```

Comment:
switch the logic to avoid the nesting. if the conditions are not met we return NULL, otherwise we continue